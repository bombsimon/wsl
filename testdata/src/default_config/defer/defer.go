package testpkg

import (
	"fmt"
	"os"
	"sync"
)

type T int

func Fn() T {
	return T(0)
}

func (*T) Close() {}

func Fn() {}

func fn() {
	a := Fn()
	b := Fn()

	defer a.Close()
	defer b.Close()

	a = Fn()
	defer a.Close()

	b = Fn()
	defer b.Close()

	a = Fn()
	defer a.Close()
	b = Fn() // want "missing whitespace decreases readability"
	defer b.Close()

	a = Fn()
	b = Fn()
	defer a.Close() // want "missing whitespace decreases readability"
	defer b.Close()

	m := sync.Mutex{}

	m.Lock()
	defer m.Unlock()

	c := true
	defer func(b bool) { // want "missing whitespace decreases readability"
		fmt.Printf("%v", b)
	}()

	_ = c
}

func fn2() {
	a := 1
	b := Fn() // want "missing whitespace decreases readability"
	defer b.Close()

	_ = a
}

func fn3() {
	f, err := os.Open("x")
	if err != nil {
		panic(err)
	}
	defer f.Close()
}
