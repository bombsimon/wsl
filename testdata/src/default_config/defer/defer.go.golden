package testpkg

import "fmt"

func fn1() {
	undoMaxProcs, err := maxprocsSet()
	if err != nil {
		return fmt.Errorf("failed to set GOMAXPROCS, err: %w", err)
	}
	defer undoMaxProcs()

	callback, x := getCb()
	if x != b {
		return
	}
	defer callback()

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	db, err := OpenDB()
	requireNoError(t, err)
	defer db.Close()

	tx := BeginTx(db)
	defer func() {
		EndTx(tx, err)
	}()

	thingOne := getOne()
	thingTwo := getTwo()
	defer thingOne.Close()
}

func fn2() {
	thingOne := getOne()

	thingTwo := getTwo()
	defer thingTwo.Close() // want "only one cuddle assignment allowed before defer statement"

	someF, err := getF()
	if err != nil {
		return err
	}

	someF() // want "expressions should not be cuddled with block"

	thingOne := getOne()
	defer thingOne.Close()

	thingTwo := getTwo()   // want "assignments should only be cuddled with other assignments"
	defer thingTwo.Close() // want "only one cuddle assignment allowed before defer statement"

	thingOne := getOne()
	thingTwo := getTwo()

	defer thingOne.Close() // want "only one cuddle assignment allowed before defer statement"
	defer thingTwo.Close()
}
